cmd/dependencyverifier/dependencyverifier.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+)
+
```

Comment:
thinking through the workflow of identifying a module we don't want, then gradually eliminating it from being vendored, from referencing it directly, and from referencing it indirectly, something like this might be clearer as input config:

```go
type Unwanted struct {
	// things we want to stop referencing
	Spec UnwantedSpec `json:"spec"`
	// status of our unwanted dependencies
	Status UnwantedStatus `json:"status"`
}
type UnwantedSpec struct {
	// module names/patterns of root modules whose dependencies should be considered direct references
	RootModules []string  `json:"rootModules"`
	// module names we don't want to depend on, mapped to an optional message about why
	UnwantedModules map[string]string  `json:"unwantedModules"`
}
type UnwantedStatus struct {
	// unwanted modules we still vendor, based on vendor/modules.txt content.
	// eliminating things from this list is good.
	Vendored []string  `json:"vendored"`
	// unwanted modules we still directly reference from spec.roots, based on `go mod graph` content.
	// eliminating things from this list is good.
	DirectReferences []string  `json:"directReferences"`
	// unwanted modules indirectly referenced from modules other than spec.roots, based on `go mod graph` content.
	// eliminating things from this list is good, but usually requires working with upstreams to do so.
	IndirectReferences []string `json:"indirectReferences"`
}
```

you can get syntax checking for free from the json decoder like this:
```
       config := &Unwanted{}
       decoder:=json.NewDecoder(bytes.NewBuffer([]byte(dependencies)))
       decoder.DisallowUnknownFields()
       if err := decoder.Decode(config);err != nil {
               log.Fatalf("Error reading dependencies file %s: %s", dependenciesJSONPath, err)
       }
```

then we cleanly separate intent ("here are our root modules, here are our unwanted modules") from status ("still vendored, still referenced from root modules, still referenced indirectly")

our input file would look like:
```yaml
{ 
  "spec": {
    "rootModules": [
      "k8s.io/*"
    ],
    "unwantedModules": {
      "github.com/go-kit/kit": "lots of transitive dependencies, see https://github.com/prometheus/common/issues/255",
      "github.com/go-openapi/analysis": "use k8s.io/kube-openapi/pkg/validation/spec",
      ...
      "github.com/hashicorp/hcl": "",
      "github.com/spf13/viper": "refer to #102598"
    }
  },
  "status": {
    "vendored": [
      ... unwantedModules still showing as vendored in modules.txt
    ],
    "directReferences": [
      ... unwantedModules still showing as dependencies of spec.rootModules in go mod graph output
    ],
    "indirectReferences": [
      ... unwantedModules still showing as dependencies of modules other than spec.rootModules in go mod graph output
    ]
  }
}
```

Then the tool can compute status, do a deepequal check against the input's status and if they differ, dump the expected status

If we start in this direction in this PR just with spec.unwantedModules / status.indirectReferences, the rootModules/vendored/directReferences bits can be added in follow-ups

Reply:
Updated with the structs and currently only update `status.indirectReferences` according to `UnwantedModules`.